### ISSUE TITLE
Quiz and Judge. Probs normalization.

### ISSUE THREAD

**Issue by HayrapetyanZhirayr** [open-sciencelab/GraphGen#66]
In the quiz and judge section, particularly in the judge part, I’ve noticed that the yes/no probabilities are not normalized.

For example, suppose the trainee model predicts Token("yes", p=0.1) while the ground truth is "no". In this case, we assume that the model’s probability for "no" is p("no") = 0.9, and we compute the loss as loss = -log(0.9). This results in a small loss, indicating that the classifier assigned a high probability (0.9) to the correct class, but actually the final verdict was incorrect. 

Could you please comment on this behavior?

**Comment by ChenZiHong-Gavin** [comment]
> In the quiz and judge section, particularly in the judge part, I’ve noticed that the yes/no probabilities are not normalized.
> 
> For example, suppose the trainee model predicts Token("yes", p=0.1) while the ground truth is "no". In this case, we assume that the model’s probability for "no" is p("no") = 0.9, and we compute the loss as loss = -log(0.9). This results in a small loss, indicating that the classifier assigned a high probability (0.9) to the correct class, but actually the final verdict was incorrect.
> 
> Could you please comment on this behavior?

Thanks for your attention.
The code for calculating yes/no loss lies [here](https://github.com/open-sciencelab/GraphGen/blob/main/graphgen/utils/calculate_confidence.py).

```python
def yes_no_loss_entropy(
    tokens_list: List[List[Token]], ground_truth: List[str]
) -> float:
    """Calculate the loss for yes/no question using entropy."""
    losses = []
    for i, tokens in enumerate(tokens_list):
        token = tokens[0]
        assert token.text.lower() in ["yes", "no"]
        if token.text == ground_truth[i]:
            losses.append(-math.log(token.prob))
        else:
            losses.append(-math.log(1 - token.prob))
    return sum(losses) / len(losses)
```

When the model stores only `Token(text="yes", prob=0.1)`, it is not saying “the answer is yes”.
It is only saying “I assign probability 0.1 to the token ‘yes’”.
Because the task is binary, the missing probability mass 1 – 0.1 = 0.9 is automatically assigned to the only other legal token, “no”.
So `Token(text="yes", prob=0.1)` equals `Token(text="no", prob=0.9)`.
Hence `P("no") = 0.9`.
The ground-truth is “no”, so the classifier’s most-probable class is already correct.
The loss `−log P("no") = −log 0.9 ≈ 0.105` is small, which is exactly what we want—because loss normally measures the divergence from the correct answer, and a lower loss means the prediction is closer to being right.

Therefore the code is correct; You may exam it with the following code:
```
from typing import List
from dataclasses import dataclass
import math

@dataclass
class Token:
    text: str
    prob: float

def yes_no_loss_entropy(
    tokens_list: List[List[Token]], ground_truth: List[str]
) -> float:
    """Calculate the loss for yes/no question using entropy."""
    losses = []
    for i, tokens in enumerate(tokens_list):
        token = tokens[0]
        assert token.text.lower() in ["yes", "no"]
        if token.text == ground_truth[i]:
            losses.append(-math.log(token.prob))
        else:
            losses.append(-math.log(1 - token.prob))
    return sum(losses) / len(losses)

if __name__ == "__main__":
    # Example usage
    tokens_list = [
        [Token(text="yes", prob=0.1)],
    ]
    ground_truth = ["no"]
    loss = yes_no_loss_entropy(tokens_list, ground_truth)
    print(f"Calculated loss: {loss}")
```

Please feel free to comment if you have any further questions.
